_posts/2020-03-11-serverless-express-sls-client.md

```diff
@@ -0,0 +1,148 @@
+        const { credentials: {
+            accessKeyId, secretAccessKey
+        }} = new Config();
+
+        const url = new URL(config.url);
+        const data = config.data ? config.data : "";
+
+        const headers: { [key: string]: string }[] = [
+            { 'X-Amz-Content-Sha256': hash(data) },
+        ];
+
+        if (!new Set(['OPTIONS', 'GET']).has(config.method.toUpperCase())) {
+            headers.push({ 'Content-Type': config.headers['Content-Type'] })
+        }
+    
+        const req = aws4.sign({
+            service: 'execute-api',
+            region: 'ap-southeast-2',
+            method: config.method.toUpperCase(),
+            path: `${url.pathname}${url.search}`,
+            headers: Object.assign({}, ...headers),
+            body: data,
+            host: url.host
+        }, { accessKeyId, secretAccessKey });
+
+        config.headers = req.headers;
+        return config
+    })
+
+    const api = new DefaultApi({}, process.env["ENDPOINT"], instance);
+```
+
+Assuming your API Gateway endpoint is loaded, this can now be used to sign requests that require IAM authorization. Assuming that the credentials that have been used have access to the invoke the relevant API Gateway endpoint.
+
+# A Comparison with JWT Authorizers
+
+It makes sense to talk about the difference between this method and JWT's, given that JWT support is available in HTTP APIs for API Gateway and IAM authorization isn't (it's limited to REST APIs). I don't think this means that AWS is abandoning IAM authorization for API Gateway - JWT's are extremely popular and every customer was implementing their own JWT Authorizer with Lambda. I think IAM authorization has several advantages over JWTs.
+
+- It provides a different signature per request, thereby providing a way to ensure the request isn't tampered with.
+- The secret is not exposed in the request, thereby limiting the opportunities to expose the secret, either via a man-in-the-middle attack or similar vector.
+- Because the request is tied to an IAM entity, all the powers of IAM are available to determine whether the caller is allowed to perform a specific action.
+
+The downside is that this method is limited to AWS. If you were to port the API to another provider you would need to implement another authorization method. You also need to aquire the credentials in the first place. Embedding credentials in an application is usually not a good idea - so most applications will use cognito federation, to enable clients to exchange a JWT from an Identity Provider for AWS Access Tokens. So even if you do decide to use IAM Authorization for your API, for a public API you will likely still end up with a JWT somewhere. I personally believe that it is worth, given how powerful and flexible IAM policies can be. The addition of session tags and scope propogation to Cognito would also offer flexible ways to control access to protected resources, but we might be waiting awhile.
```

Comment:
>I personally believe that it is worth, given how powerful and flexible IAM policies can be.

This sentence reads strange. The one preceeding it doesn't help make any more sense either.